Create a script for rotating secrets (#287)
gardener/ci-infra: README.md

You can test your TestGrid configuration locally with the `./hack/check-testgrid-config.sh`. Please open a PR for `ci-infra` repository for your new configuration. When it is merged the new configuration will be pushed to `gs://gardener-prow/testgrid/config` automatically and your jobs will become visible at [testgrid.k8s.io/gardener](https://testgrid.k8s.io/gardener) soon.

## Combined kubeconfig for prow clusters and Gardener project

The scripts from this repository rely on a combined `kubeconfig`. It contains two contexts for the prow clusters `gardener-prow-trusted`, `gardener-prow-build` and one for the Gardener project the clusters are created in.
Please setup your local kubeconfig file by using the `hack/setup-prow-kubeconfig.sh` script. Afterwards, you find it here:
```bash
export KUBECONFIG=~/.gardener-prow/kubeconfig/kubeconfig--gardener--prow-combined.yaml
```
The kubeconfig contains absolute paths. Thus, it won't work anymore, if you move it to a different location.

## How to setup

> The following commands assume you are using the combined `kubeconfig` which introduced in the previous section. When you create new clusters the configuration of `gardener-prow-trusted`, `gardener-prow-build`  contextes will be incomplete in the beginning. They are completed in step 2 when the clusters are created.

Reviewer: ```suggestion
The following commands assume you are using the combined `kubeconfig` generated in the previous section. When you create new clusters the configuration of `gardener-prow-trusted`, `gardener-prow-build`  contexts will be incomplete in the beginning. They are completed in step 2 when the clusters have been created.
```

```yaml
apiVersion: v1
clusters:
- cluster:
    server: https://api.live.gardener.cloud.sap
  name: garden-garden-ci
- cluster:
    certificate-authority-data: <<Get from your gardener project>>
    server: <<Get from your gardener project>>
  name: shoot--garden-ci--prow
- cluster:
    certificate-authority-data: <<Get from your gardener project>>
    server: <<Get from your gardener project>>
  name: shoot--garden-ci--prow-work
contexts:
- context:
    cluster: garden-garden-ci
    namespace: garden-garden-ci
    user: oidc-login
  name: garden-garden-ci
- context:
    cluster: shoot--garden-ci--prow-work
    user: shoot--garden-ci--prow-work-token
  name: gardener-prow-build
- context:
    cluster: shoot--garden-ci--prow
    user: shoot--garden-ci--prow-token
  name: gardener-prow-trusted
current-context: gardener-prow-trusted
kind: Config
users:
- name: oidc-login
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - <<Get from your gardener project>>
      command: kubectl
- name: shoot--garden-ci--prow-token
  user:
    token: <<Get from your gardener project>>
- name: shoot--garden-ci--prow-work-token
  user:

>     token: <<Get from your gardener project>>

Reviewer: Instead of using a non-personalized token, we could use `gardenlogin`, e.g. something like what is generated by `gardenctl`:
```yaml
clusters:
- cluster:
    certificate-authority-data: ...
    extensions:
    - extension:
        gardenClusterIdentity: sap-landscape-live
        shootRef:
          name: prow-work
          namespace: garden-garden-ci
      name: client.authentication.k8s.io/exec
    server: ...
  name: garden-garden-ci--prow-work-external
users:
- name: garden-garden-ci--prow-work
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - gardenlogin
      - get-client-certificate
      command: kubectl
      env: null
      interactiveMode: IfAvailable
      provideClusterInfo: true
```

```

## How to setup

The following commands assume you are using the combined `kubeconfig` which introduced in the previous section. When you create new clusters the configuration of `gardener-prow-trusted`, `gardener-prow-build`  contextes will be incomplete in the beginning. They are completed in step 2 when the clusters are created.

1. Create the prow cluster and prow workload cluster.
   ```bash
   kubectl config use-context garden-garden-ci
   kubectl apply -f config/clusters/shoot.yaml
   kubectl apply -f config/clusters/shoot-workload.yaml
   ```
1. Complete your combined kubeconfig with the data of the clusters created in the previous step
1. Create the `prow` namespace in the prow cluster:
   ```bash
   kubectl config use-context gardener-prow-trusted
   kubectl apply --server-side=true -f config/prow/cluster/prow_namespace.yaml
   ```
1. Create the `test-pods` namespace in the workload/build cluster:
   ```bash
   kubectl config use-context gardener-prow-build

>    kubectl apply --server-side=true -f config/prow/cluster/build

Reviewer: ```suggestion
   kubectl apply --server-side=true -f config/prow/cluster/base/test-pods_namespace.yaml
```
